src/control/cmd/daos_agent/security_rpc_test.go

```diff
@@ -188,3 +197,125 @@ func TestAgentSecurityModule_RequestCreds_BadUid(t *testing.T) {
 
 	expectCredResp(t, respBytes, int32(daos.MiscError), false)
 }
+
+func TestAgent_SecurityRPC_getCredential(t *testing.T) {
+	type response struct {
+		cred *auth.Credential
+		err  error
+	}
+	testCred := &auth.Credential{
+		Token:  &auth.Token{Flavor: auth.Flavor_AUTH_SYS, Data: []byte("test-token")},
+		Origin: "test-origin",
+	}
+	miscErrBytes, err := proto.Marshal(
+		&auth.GetCredResp{
+			Status: int32(daos.MiscError),
+		},
+	)
+	if err != nil {
+		t.Fatalf("Couldn't marshal misc error: %v", err)
+	}
+	successBytes, err := proto.Marshal(
+		&auth.GetCredResp{
+			Status: 0,
+			Cred:   testCred,
+		},
+	)
+	if err != nil {
+		t.Fatalf("Couldn't marshal success: %v", err)
+	}
+
+	for name, tc := range map[string]struct {
+		secCfg    *securityConfig
+		responses []response
+		expBytes  []byte
+		expErr    error
+	}{
+		"lookup miss": {
+			secCfg: defaultTestSecurityConfig(),
+			responses: []response{
+				{
+					cred: nil,
+					err:  user.UnknownUserIdError(unix.Getuid()),
+				},
+			},
+			expBytes: miscErrBytes,
+		},
+		"lookup OK": {
+			secCfg: func() *securityConfig {
+				cfg := defaultTestSecurityConfig()
+				cfg.credentials.ClientUserMap = security.ClientUserMap{
+					uint32(unix.Getuid()): &security.MappedClientUser{
+						User: "test-user",
+					},
+				}
+				return cfg
+			}(),
+			responses: []response{
+				{
+					cred: nil,
+					err:  user.UnknownUserIdError(unix.Getuid()),
+				},
+				{
+					cred: testCred,
+					err:  nil,
+				},
+			},
+			expBytes: successBytes,
+		},
+		"lookup OK, but retried request fails": {
+			secCfg: func() *securityConfig {
+				cfg := defaultTestSecurityConfig()
+				cfg.credentials.ClientUserMap = security.ClientUserMap{
+					uint32(unix.Getuid()): &security.MappedClientUser{
+						User: "test-user",
+					},
+				}
+				return cfg
+			}(),
+			responses: []response{
+				{
+					cred: nil,
+					err:  user.UnknownUserIdError(unix.Getuid()),
+				},
+				{
+					cred: nil,
+					err:  errors.New("oops"),
+				},
+			},
+			expBytes: miscErrBytes,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			log, buf := logging.NewTestLogger(t.Name())
+			defer test.ShowBufferOnFailure(t, buf)
+
+			// Set up a real unix socket so we can make a real connection
+			conn, cleanup := setupTestUnixConn(t)
+			defer cleanup()
+
+			mod := NewSecurityModule(log, tc.secCfg)
+			mod.signCredential = func() func(req *auth.CredentialRequest) (*auth.Credential, error) {
```

Comment:
Why `func() sig {...} ()` Rather than just `sig {...}` here and specify `var idx int` at top-level inside `t.Run()`?

Reply:
I've changed the code to use the `credSignerFn` type in the return signature to make it clearer that the closure is generating a function. The index variable is scoped to the generator closure and the generated closure.